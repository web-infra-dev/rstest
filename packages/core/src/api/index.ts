import { createRunner } from '../runner';
import type {
  Rstest,
  RstestExpect,
  RunnerHooks,
  TestCase,
  TestSummaryResult,
  WorkerState,
} from '../types';
import { GLOBAL_EXPECT, createExpect } from './expect';

export const createRstestRuntime = (
  workerState: WorkerState,
): {
  runner: {
    runTest: (
      testPath: string,
      hooks: RunnerHooks,
    ) => Promise<TestSummaryResult>;
    getCurrentTest: () => TestCase | undefined;
  };
  api: Rstest;
} => {
  const { runner, api: runnerAPI } = createRunner({ workerState });

  const expect: RstestExpect = createExpect({
    workerState,
    getCurrentTest: () => runner.getCurrentTest(),
  });

  Object.defineProperty(globalThis, GLOBAL_EXPECT, {
    value: expect,
    writable: true,
    configurable: true,
  });

  return {
    runner,
    api: {
      expect,
      ...runnerAPI,
    },
  };
};
