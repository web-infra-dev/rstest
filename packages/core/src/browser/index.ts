import type { Rstest } from '../core/rstest';
import { runBrowserController } from './hostController';

export async function runBrowserTests(context: Rstest): Promise<void> {
  await runBrowserController(context);
}
