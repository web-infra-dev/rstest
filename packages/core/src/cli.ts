export { runCLI } from './cli/index';
