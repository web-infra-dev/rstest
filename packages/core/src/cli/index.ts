import { logger } from '../utils/logger';
import { setupCommands } from './commands';
import { prepareCli } from './prepare';

export async function runCLI(): Promise<void> {
  prepareCli();

  try {
    setupCommands();
  } catch (err) {
    logger.error('Failed to start Rstest CLI.');
    logger.error(err);
  }
}
