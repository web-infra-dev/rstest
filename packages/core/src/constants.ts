import type { Rstest } from './types';

export const DEFAULT_CONFIG_NAME = 'rstest.config';

export const DEFAULT_CONFIG_EXTENSIONS = [
  '.js',
  '.ts',
  '.mjs',
  '.mts',
  '.cjs',
  '.cts',
] as const;

export const globalApis: (keyof Rstest)[] = [
  'test',
  'describe',
  'it',
  'expect',
];
