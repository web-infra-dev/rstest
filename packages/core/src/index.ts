export { defineConfig } from './config';

export * from './api/public';

// TODO: Move below to @rstest/node exports point.
export { createRstest } from './core';
export type { RstestConfig } from './types';
