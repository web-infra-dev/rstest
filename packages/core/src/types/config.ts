import type { BuiltInReporterNames, Reporter } from './reporter';

export type RstestPoolType = 'forks';

export type RstestPoolOptions = {
  /** Pool used to run tests in. */
  type: RstestPoolType;
  /** Maximum number or percentage of workers to run tests in. */
  maxWorkers?: number | string;
  /** Minimum number or percentage of workers to run tests in. */
  minWorkers?: number | string;
  /** Pass additional arguments to node process in the child processes. */
  execArgv?: string[];
};

export interface RstestConfig {
  /**
   * Project root
   *
   * @default process.cwd()
   */
  root?: string;
  /**
   * Project name
   *
   * @default rstest
   */
  name?: string;
  /**
   * A list of glob patterns that match your test files.
   *
   * @default ['**\/*.{test,spec}.?(c|m)[jt]s?(x)']
   */
  include?: string[];
  /**
   * A list of glob patterns that should be excluded from your test files.
   *
   * @default ['**\/node_modules/**', '**\/dist/**']
   */
  exclude?: string[];
  /**
   * Path to setup files. They will be run before each test file.
   */
  setupFiles?: string[] | string;
  /**
   * Allows the test suite to pass when no files are found.
   *
   * @default false
   */
  passWithNoTests?: boolean;
  /**
   * Pool used to run tests in.
   */
  pool?: RstestPoolType | RstestPoolOptions;
  /**
   * Run tests in an isolated environment
   *
   * @default true
   */
  isolate?: boolean;
  /**
   * Provide global APIs
   *
   * @default false
   */
  globals?: boolean;
  /**
   * Update snapshot files. Will update all changed snapshots and delete obsolete ones.
   *
   * @default false
   */
  update?: boolean;
  /**
   * Custom reporter for output.
   * @default ['default']
   */
  reporters?:
    | Reporter
    | BuiltInReporterNames
    | (Reporter | BuiltInReporterNames)[];
  /**
   * Run only tests with a name that matches the regex.
   */
  testNamePattern?: string | RegExp;

  /**
   * Timeout of a test in milliseconds.
   * @default 5000
   */
  testTimeout?: number;

  /**
   * Automatically clear mock calls, instances, contexts and results before every test.
   * @default false
   */
  clearMocks?: boolean;
  /**
   * Automatically reset mock state before every test.
   * @default false
   */
  resetMocks?: boolean;
  /**
   * Automatically restore mock state and implementation before every test.
   * @default false
   */
  restoreMocks?: boolean;
  /**
   * Restores all global variables that were changed with `rstest.stubGlobal` before every test.
   * @default false
   */
  unstubGlobals?: boolean;
  /**
   * Restores all `process.env` values that were changed with `rstest.stubEnv` before every test.
   * @default false
   */
  unstubEnvs?: boolean;
}

export type NormalizedConfig = Required<
  Omit<RstestConfig, 'pool' | 'setupFiles' | 'testNamePattern'>
> & {
  pool: RstestPoolOptions;
  setupFiles?: string[] | string;
  testNamePattern?: RstestConfig['testNamePattern'];
};
