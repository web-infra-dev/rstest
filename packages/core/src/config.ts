import fs from 'node:fs';
import { dirname, isAbsolute, join } from 'node:path';
import {
  type LoadConfigOptions,
  loadConfig as loadRsbuildConfig,
  mergeRsbuildConfig,
} from '@rsbuild/core';
import { DEFAULT_CONFIG_EXTENSIONS, DEFAULT_CONFIG_NAME } from './constants';
import type { NormalizedConfig, RstestConfig } from './types';
import { color } from './utils/helper';
import { logger } from './utils/logger';

export type RstestConfigAsyncFn = () => Promise<RstestConfig>;

export type RstestConfigSyncFn = () => RstestConfig;

export type RstestConfigExport =
  | RstestConfig
  | RstestConfigSyncFn
  | RstestConfigAsyncFn;

/**
 * This function helps you to autocomplete configuration types.
 * It accepts a Rsbuild config object, or a function that returns a config.
 */
export function defineConfig(config: RstestConfig): RstestConfig;
export function defineConfig(config: RstestConfigSyncFn): RstestConfigSyncFn;
export function defineConfig(config: RstestConfigAsyncFn): RstestConfigAsyncFn;
export function defineConfig(config: RstestConfigExport): RstestConfigExport;
export function defineConfig(config: RstestConfigExport) {
  return config;
}

const findConfig = (basePath: string): string | undefined => {
  return DEFAULT_CONFIG_EXTENSIONS.map((ext) => basePath + ext).find(
    fs.existsSync,
  );
};

const resolveConfigPath = (root: string, customConfig?: string) => {
  if (customConfig) {
    const customConfigPath = isAbsolute(customConfig)
      ? customConfig
      : join(root, customConfig);
    if (fs.existsSync(customConfigPath)) {
      return customConfigPath;
    }
    logger.warn(`Cannot find config file: ${color.dim(customConfigPath)}\n`);
  }

  const configFilePath = findConfig(join(root, DEFAULT_CONFIG_NAME));

  if (configFilePath) {
    return configFilePath;
  }

  return null;
};

export async function loadConfig({
  cwd = process.cwd(),
  path,
  envMode,
  configLoader,
}: {
  cwd?: string;
  path?: string;
  envMode?: string;
  configLoader?: LoadConfigOptions['loader'];
}): Promise<{
  content: RstestConfig;
  filePath: string | null;
}> {
  const configFilePath = resolveConfigPath(cwd, path);

  if (!configFilePath) {
    logger.debug('no config file found');
    return {
      content: {},
      filePath: configFilePath,
    };
  }

  const { content } = await loadRsbuildConfig({
    cwd: dirname(configFilePath),
    path: configFilePath,
    envMode,
    loader: configLoader,
  });

  return { content: content as RstestConfig, filePath: configFilePath };
}

export const mergeRstestConfig = (...configs: RstestConfig[]): RstestConfig =>
  mergeRsbuildConfig<RstestConfig>(...configs);

const createDefaultConfig = (): NormalizedConfig => ({
  root: process.cwd(),
  name: 'rstest',
  include: ['**/*.{test,spec}.?(c|m)[jt]s?(x)'],
  exclude: ['**/node_modules/**', '**/dist/**'],
});

export const withDefaultConfig = (config: RstestConfig): NormalizedConfig => {
  const merged = mergeRstestConfig(createDefaultConfig(), config);

  // The following configurations need overrides
  merged.include = config.include || merged.include;
  merged.exclude = config.exclude || merged.exclude;

  return merged as NormalizedConfig;
};
