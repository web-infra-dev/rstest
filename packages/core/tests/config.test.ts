import { mergeRstestConfig, withDefaultConfig } from '../src/config';

describe('mergeRstestConfig', () => {
  it('should merge config correctly with default config', () => {
    const merged = withDefaultConfig({
      include: ['tests/**/*.test.ts'],
      root: __dirname,
      exclude: ['**/aa/**'],
      setupFiles: ['./setup.ts'],
      reporters: ['verbose'],
    });

    expect(merged).toMatchSnapshot();
  });

  it('should merge exclude correctly', () => {
    expect(
      mergeRstestConfig(
        {
          exclude: ['**/node_modules/**'],
        },
        {
          exclude: {
            patterns: ['**/dist/**'],
            override: true,
          },
        },
      ),
    ).toEqual({
      exclude: {
        patterns: ['**/dist/**'],
      },
    });

    expect(
      mergeRstestConfig(
        {
          exclude: ['**/node_modules/**'],
        },
        {
          exclude: {
            patterns: ['**/dist/**'],
            override: false,
          },
        },
      ),
    ).toEqual({
      exclude: {
        patterns: ['**/node_modules/**', '**/dist/**'],
        override: false,
      },
    });

    expect(
      mergeRstestConfig(
        {
          exclude: {
            patterns: ['**/dist/**'],
            override: false,
          },
        },
        {
          exclude: ['**/node_modules/**'],
        },
        {
          exclude: {
            patterns: ['**/aa/**'],
            override: true,
          },
        },
      ),
    ).toEqual({
      exclude: {
        patterns: ['**/aa/**'],
      },
    });
  });
});
