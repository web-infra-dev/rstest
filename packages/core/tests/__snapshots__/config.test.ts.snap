// Rstest Snapshot v1

exports[`mergeRstestConfig > should merge config correctly with default config 1`] = `
{
  "clearMocks": false,
  "coverage": {
    "clean": true,
    "enabled": false,
    "exclude": [
      "**/node_modules/**",
      "**/dist/**",
      "**/test/**",
      "**/__tests__/**",
      "**/__mocks__/**",
      "**/*.d.ts",
      "**/*.{test,spec}.[jt]s",
      "**/*.{test,spec}.[c|m][jt]s",
      "**/*.{test,spec}.[jt]sx",
      "**/*.{test,spec}.[c|m][jt]sx",
    ],
    "provider": "istanbul",
    "reporters": [
      "text",
      "html",
      "clover",
      "json",
    ],
    "reportsDirectory": "<ROOT>/packages/core/tests/coverage",
  },
  "disableConsoleIntercept": false,
  "exclude": {
    "override": false,
    "patterns": [
      "**/node_modules/**",
      "**/dist/**",
      "**/.{idea,git,cache,output,temp}/**",
      "**/aa/**",
      "**/dist/.rstest-temp",
    ],
  },
  "globals": false,
  "hookTimeout": 10000,
  "include": [
    "tests/**/*.test.ts",
  ],
  "includeSource": [],
  "isolate": true,
  "maxConcurrency": 5,
  "name": "rstest",
  "passWithNoTests": false,
  "pool": {
    "type": "forks",
  },
  "printConsoleTrace": false,
  "reporters": [
    "verbose",
  ],
  "resetMocks": false,
  "restoreMocks": false,
  "retry": 0,
  "root": "<ROOT>/packages/core/tests",
  "setupFiles": [
    "./setup.ts",
  ],
  "slowTestThreshold": 300,
  "testEnvironment": "node",
  "testTimeout": 5000,
  "unstubEnvs": false,
  "unstubGlobals": false,
  "update": false,
}
`;
